File: sources/CPP/include/Ers/Math/RandomGeneratorOwned.cpp
#include "RandomGeneratorOwned.h"

#include <utility>

#include "Ers/Api.h"

namespace Ers
{
    /// @brief Creates a random number generator with the given pointer to the RandomGenerator in the core.
    /// @param coreInstance The pointer to the instance of an existing RandomGenerator in the core.
    /// @param owned Whether the RandomGenerator is owned and thus may be destroyed on deconstruction.
    RandomGeneratorOwned::RandomGeneratorOwned(void* coreInstance) :
        coreRandomGeneratorInstance(coreInstance)
    {
    }

    RandomGeneratorOwned::~RandomGeneratorOwned()
    {
        if (coreRandomGeneratorInstance == nullptr)
            return;

        ersAPIFunctionPointers.ERS_Random_Generator_Destroy(coreRandomGeneratorInstance);
    }

    RandomGeneratorOwned::RandomGeneratorOwned(RandomGeneratorOwned&& other) :
        coreRandomGeneratorInstance(std::exchange(other.coreRandomGeneratorInstance, nullptr))
    {
    }

    RandomGeneratorOwned& RandomGeneratorOwned::operator=(RandomGeneratorOwned&& other)
    {
        coreRandomGeneratorInstance = std::exchange(other.coreRandomGeneratorInstance, nullptr);
        return *this;
    }

    /// @brief Create a new owned Wichmann Hill pseudo-random number generator.
    /// @return The created random number generator instance.
    RandomGeneratorOwned RandomGeneratorOwned::CreateWichmannHill()
    {
        RandomGeneratorOwned generator = RandomGeneratorOwned(ersAPIFunctionPointers.ERS_Random_Generator_WichmannHill_Create());
        return generator;
    }

    /// @brief Create a new owned Mersenne Twister pseudo-random number generator.
    /// @return The created random number generator instance.
    RandomGeneratorOwned RandomGeneratorOwned::CreateMersenneTwister()
    {
        RandomGeneratorOwned generator = RandomGeneratorOwned(ersAPIFunctionPointers.ERS_Random_Generator_MersenneTwister_Create());
        return generator;
    }

    /// @brief Reset the seed of the random number generator.
    void RandomGeneratorOwned::Reset()
    {
        ersAPIFunctionPointers.ERS_Random_Generator_Reset(coreRandomGeneratorInstance);
    }

    /// @brief Set the set of the random number generator.
    /// @param seed The seed to set.
    void RandomGeneratorOwned::SetSeed(size_t seed)
    {
        ersAPIFunctionPointers.ERS_Random_Generator_Set_Seed(coreRandomGeneratorInstance, seed);
    }

    /// @brief Sample a random value from the random number generator.
    /// @return The sampled value.
    double RandomGeneratorOwned::Sample()
    {
        return ersAPIFunctionPointers.ERS_Random_Generator_Sample(coreRandomGeneratorInstance);
    }

    Ers::RandomGeneratorType RandomGeneratorOwned::GetRandomGeneratorType()
    {
        return static_cast<Ers::RandomGeneratorType>(ersAPIFunctionPointers.ERS_Random_Generator_GetType(coreRandomGeneratorInstance));
    }
} // namespace Ers

File: sources/CSharp/src/Ers/Math/RandomGeneratorOwned.cs
using Ers.Engine;

namespace Ers
{
    /// <summary>
    /// A random number generator that is not owned by user-code, not the ERS core.
    /// Therefore, it can be created and destroyed at will.
    /// </summary>
    public class RandomGeneratorOwned : IDisposable
    {
        private IntPtr coreRandomGeneratorInstance;

        internal RandomGeneratorOwned(IntPtr coreInstance) { coreRandomGeneratorInstance = coreInstance; }

        /// <summary>
        /// Destroy the random generator.
        /// </summary>
        public void Dispose()
        {
            if (coreRandomGeneratorInstance != IntPtr.Zero)
                return;

            ErsEngine.ERS_Random_Generator_Destroy(coreRandomGeneratorInstance);
            coreRandomGeneratorInstance = IntPtr.Zero;
        }

        /// <summary>
        /// Construct a new RandomGeneratorOwned of the Whichmann-Hill type.
        /// </summary>
        /// <returns></returns>
        public static RandomGeneratorOwned CreateWichmannHill()
        {
            return new RandomGeneratorOwned(ErsEngine.ERS_Random_Generator_WichmannHill_Create());
        }

        /// <summary>
        /// Construct a new RandomGeneratorOwned of the Mersenne Twister type.
        /// </summary>
        /// <returns></returns>
        public static RandomGeneratorOwned CreateMersenneTwister()
        {
            return new RandomGeneratorOwned(ErsEngine.ERS_Random_Generator_MersenneTwister_Create());
        }

        /// <summary>
        /// Reset the random number generator.
        /// </summary>
        public void Reset() => ErsEngine.ERS_Random_Generator_Reset(coreRandomGeneratorInstance);

        /// <summary>
        /// Sample a random number between 0.0 and 1.0.
        /// </summary>
        /// <returns></returns>
        public double Sample() => ErsEngine.ERS_Random_Generator_Sample(coreRandomGeneratorInstance);

        /// <summary>
        /// Set the seed of the random number generator.
        /// </summary>
        /// <param name="seed">The new seed.</param>
        public void SetSeed(nuint seed) => ErsEngine.ERS_Random_Generator_Set_Seed(coreRandomGeneratorInstance, seed);

        /// <summary>
        /// Get the type of random number generator. See <see cref="RandomGeneratorType"/>.
        /// </summary>
        /// <returns></returns>
        public RandomGeneratorType GetRandomGeneratorType()
        {
            return (RandomGeneratorType)ErsEngine.ERS_Random_Generator_GetType(coreRandomGeneratorInstance);
        }
    }
}
